src/rust_api.rs

```diff
@@ -0,0 +1,563 @@
+                    self.pending_data = Some(data);
+                } else {
+                    let _ = mem::take(&mut data);
+                }
+
+                return Err(err);
+            }
+
+            if data.data.as_ref().is_some_and(|d| d.len() > 0) {
+                self.pending_data = Some(data);
+                return Err(Rav1dError::TryAgain);
+            }
+
+            Ok(())
+        }
+
+        /// Sends any pending data to the decoder.
+        ///
+        /// This has to be called after [`Decoder::send_data`] has returned `Err([Error::Again])` to
+        /// consume any futher pending data.
+        ///
+        /// After this returned `Ok(())` or `Err([Error::Again])` there might be decoded frames
+        /// available via [`Decoder::get_picture`].
+        pub fn send_pending_data(&mut self) -> Result<(), Rav1dError> {
+            let mut data = match self.pending_data.take() {
+                None => {
+                    return Ok(());
+                }
+                Some(data) => data,
+            };
+
+            if let Err(err) = rav1d_send_data(&self.ctx, &mut data) {
+                if matches!(err, Rav1dError::TryAgain) {
+                    self.pending_data = Some(data);
+                } else {
+                    let _ = mem::take(&mut data);
+                }
+
+                return Err(err);
+            }
+
+            if data.data.as_ref().is_some_and(|d| d.len() > 0) {
+                self.pending_data = Some(data);
+                return Err(Rav1dError::TryAgain);
+            }
+
+            Ok(())
+        }
+
+        /// Get the next decoded frame from the decoder.
+        ///
+        /// If this returns `Err([Error::Again])` then further data has to be sent to the decoder
+        /// before further decoded frames become available.
+        ///
+        /// To make most use of frame threading this function should only be called once per submitted
+        /// input frame and not until it returns `Err([Error::Again])`. Calling it in a loop should
+        /// only be done to drain all pending frames at the end.
+        pub fn get_picture(&mut self) -> Result<Picture, Rav1dError> {
+            let mut pic = Rav1dPicture::default();
+            let ret = rav1d_get_picture(&self.ctx, &mut pic);
+
+            if let Err(err) = ret {
+                Err(err)
+            } else {
+                Ok(Picture {
+                    inner: Arc::new(pic),
+                })
+            }
+        }
+
+        /// Get the decoder delay.
+        pub fn get_frame_delay(&self) -> u32 {
+            // The only fields this actually needs from Rav1dSettings are n_threads and max_frame_delay so we just pass these in directly
+
+            rav1d_get_frame_delay(&Rav1dSettings {
+                n_threads: self.n_threads,
+                max_frame_delay: self.max_frame_delay,
+                ..Default::default()
+            })
+            .unwrap() as u32
```

Comment:
You can change `rav1d_get_frame_delay` to return the `usize` directly since it's always `Ok`.